crates/wasmtime/src/component/resources.rs

```diff
@@ -0,0 +1,714 @@
+use crate::component::func::{bad_type_info, desc, LiftContext, LowerContext};
+use crate::component::matching::InstanceType;
+use crate::component::{ComponentType, Lift, Lower};
+use crate::store::{StoreId, StoreOpaque};
+use crate::{AsContextMut, StoreContextMut, Trap};
+use anyhow::{bail, Result};
+use std::any::TypeId;
+use std::fmt;
+use std::marker;
+use std::mem::MaybeUninit;
+use std::sync::atomic::{AtomicU32, Ordering::Relaxed};
+use wasmtime_environ::component::{CanonicalAbiInfo, DefinedResourceIndex, InterfaceType};
+use wasmtime_runtime::component::{ComponentInstance, InstanceFlags, ResourceTables};
+use wasmtime_runtime::{SendSyncPtr, VMFuncRef, ValRaw};
+
+/// Representation of a resource type in the component model.
+///
+/// Resources are currently always represented as 32-bit integers but they have
+/// unique types across instantiations and the host. For example instantiating
+/// the same component twice means that defined resource types in the component
+/// will all be different. Values of this type can be compared to see if
+/// resources have the same type.
+///
+/// Resource types can also be defined on the host in addition to guests. On the
+/// host resource types are tied to a `T`, an arbitrary Rust type. Two host
+/// resource types are the same if they point to the same `T`.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct ResourceType {
+    kind: ResourceTypeKind,
+}
+
+impl ResourceType {
+    /// Creates a new host resource type corresponding to `T`.
+    ///
+    /// Note that `T` is a mostly a phantom type parameter here. It does not
+    /// need to reflect the actual storage of the resource `T`. For example this
+    /// is valid:
+    ///
+    /// ```rust
+    /// use wasmtime::component::ResourceType;
+    ///
+    /// struct Foo;
+    ///
+    /// let ty = ResourceType::host::<Foo>();
+    /// ```
+    ///
+    /// A resource type of type `ResourceType::host::<T>()` will match the type
+    /// of the value produced by `Resource::<T>::new_{own,borrow}`.
+    pub fn host<T: 'static>() -> ResourceType {
+        ResourceType {
+            kind: ResourceTypeKind::Host(TypeId::of::<T>()),
+        }
+    }
+
+    pub(crate) fn guest(
+        store: StoreId,
+        instance: &ComponentInstance,
+        id: DefinedResourceIndex,
+    ) -> ResourceType {
+        ResourceType {
+            kind: ResourceTypeKind::Guest {
+                store,
+                instance: instance as *const _ as usize,
+                id,
+            },
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ResourceTypeKind {
+    Host(TypeId),
+    Guest {
+        store: StoreId,
+        // For now this is the `*mut ComponentInstance` pointer within the store
+        // that this guest corresponds to. It's used to distinguish different
+        // instantiations of the same component within the store.
+        instance: usize,
+        id: DefinedResourceIndex,
+    },
+}
```

Comment:
If a component takes two resource type imports, and is instantiated from the host with the same resource for both imports, and we get the `ResourceType` for each of those imports *from that component*, will both of those result in the same `ResourceType`? That is, even though the compiled component cannot know they are the same, at the Wasmtime API level when we do reflection on the instantiated component, do we see through everything and determine they are the same? I think we should, and I think we should test this if we don't already.

Reply:
Correct! Added a test as well 👍